docs/enhancements/2025-07-18-generic-controller.md

```diff
@@ -0,0 +1,576 @@
+                          route: { cluster: foo-svc }
+
+                    - name: wildcard
+                      domains: ["*.example.com"]
+                      routes:
+                        - match:
+                            prefix: "/"
+                            headers:
+                              - name: env
+                                exact_match: "canary"
+                          route: { cluster: bar-svc-canary }
+
+                        - match: { prefix: "/" }
+                          route: { cluster: bar-svc }
+                http_filters:
+                  - name: envoy.filters.http.router
+
+  clusters:
+    - name: foo-svc
+      type: strict_dns
+      load_assignment:
+        cluster_name: foo-svc
+        endpoints:
+          - lb_endpoints:
+              - endpoint: { address: { socket_address: { address: 10.0.0.10, port_value: 8080 } } }
+              - endpoint: { address: { socket_address: { address: 10.0.0.20, port_value: 8080 } } }
+
+    - name: bar-svc-canary
+      type: strict_dns
+      load_assignment:
+        cluster_name: bar-svc-canary
+        endpoints:
+          - lb_endpoints:
+              - endpoint: { address: { socket_address: { address: 10.10.10.10, port_value: 8080 } } }
+              - endpoint: { address: { socket_address: { address: 10.10.10.20, port_value: 8080 } } }
+
+    - name: bar-svc
+      type: strict_dns
+      load_assignment:
+        cluster_name: bar-svc
+        endpoints:
+          - lb_endpoints:
+              - endpoint: { address: { socket_address: { address: 10.20.20.10, port_value: 8080 } } }
+              - endpoint: { address: { socket_address: { address: 10.20.20.20, port_value: 8080 } } }
+```
+
+As we can see, each backend definition to be used will generate a new cluster inside
+the `clusters` array.
+
+And every new route will be attached to a filter of the relative listener, that
+was previously defined on the `Gateway` definitions
+
+### Conclusions on Route reconciliation
+For route reconciliation, we can see that the same pattern happens for the 3 major
+backends:
+* For the workload IP / endpoint definitions, it will generate a specific block that
+contains just "who should be contacted" (and how, timeouts, etc)
+* For the routes, they will be attached to the listener based on the desired rules,
+and direct the traffic to the proper backends.
+
+This means that for a Route reconciliation, we need the following operations:
+* Add/Update/Delete a group of backend/upstreams (based on Endpoints/EndpointSlices)
+* Update a Listener to attach/update/detach a route, and point to a backend.
+
+# Consuming InGate as a library
+
+To make it possible for backends to consume InGate as a library, a proposal would be
+to make InGate request, during its instantiation, structures that comply with some interfaces.
+
+Given the following interfaces definitions (not final, just examples)
+
+```go
+type ListenerController interface {
```

Comment:
I know this an example, so to be clear a listenerController is just a gateway controller right?